### PR TITLE
ICMSLST-1050 Add CFS schedule template editor.

### PR DESCRIPTION
**Pull request by MattHolmes123**
Changes:
  - Add the following CFS template views:
    - CFSScheduleTemplateAddView
    - CFSScheduleTemplateCopyView
    - CFSScheduleTemplateDeleteView
    - CFSManufacturerUpdateView
    - CFSManufacturerDeleteView
  - Update main CFS template to show schedules.
  - Improve all redirects to go back to step being edited.
  - Add error handling when template data is being copied to an application.
  - Refactor and share several templates between CFS Application and CFS
    Template editor.
  - Add template Models for all equivalent CFS models.

Edit CFS Template view:
export-a-certificate_8080_cat_edit_1_step_cfs_

Edit CFS Schedule Template
export-a-certificate_8080_cat_edit_1_step_cfs-schedule_step_pk_1_

Edit Manufacturer details (shows postcode lookup working)
export-a-certificate_8080_cat_edit_1_schedule_template_1_cfs-manufacturer-update_

### CODE REVIEW

File: web/domains/cat/forms.py [hunk 1]
```diff
@@ -59,8 +67,29 @@ class Meta:
...
     class Meta:
         model = CertificateOfGoodManufacturingPracticeApplicationTemplate
         fields = copy_form_fields(EditGMPForm.Meta.fields, "contact")
         widgets = EditGMPForm.Meta.widgets
+
+
+class CertificateOfFreeSaleApplicationTemplateForm(EditCFSForm):
+    class Meta:
+        model = CertificateOfFreeSaleApplicationTemplate
+        fields = copy_form_fields(EditCFSForm.Meta.fields, "contact")
```
Comment (chris-pettinga) [inline web/domains/cat/forms.py:80]: I like this, have been bitten badly in the past by ModelForm Meta inheritance, a real pain.

File: web/domains/case/export/views.py [hunk 2]
```diff
@@ -490,6 +465,24 @@ def cfs_edit_schedule(
         return render(request, "web/domains/case/export/cfs-edit-schedule.html", context)
 
 
+def get_csf_schedule_legislation_config():
```
Comment (chris-pettinga) [inline web/domains/case/export/views.py:468]:
```suggestion
def get_csf_schedule_legislation_config() -> Dict[UUID: Dict[str: bool]]:
```

File: web/domains/cat/utils.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+            #         ptn.save()
+            #
+            #     for ingredient in ingredients_to_copy:
+            #         ingredient.pk = None
+            #         ingredient._state.adding = True
+            #         ingredient.product = product
+            #         ingredient.save()
+
+
+def _save_form(form: ModelForm, cat_pk) -> Any:
```
Comment (chris-pettinga) [inline web/domains/cat/utils.py:106]:
```suggestion
def _save_form(form: ModelForm, cat_pk: UUID) -> Any:
```

**Review by chris-pettinga — APPROVED** [review]

File: web/domains/case/export/views.py (hunk 2 above)
Comment (MattHolmes123) [reply web/domains/case/export/views.py:468]: the PK is an int.
`dict[int, dict[str, bool]]`

File: web/domains/cat/utils.py (hunk 3 above)
Comment (MattHolmes123) [reply web/domains/cat/utils.py:106]: `int` 👍

File: web/domains/cat/views.py [hunk 4]
```diff
@@ -151,45 +210,85 @@ def get_form(self, form_class=None) -> ModelForm:
...
-        return super().form_valid(form)
-
-    def get_page_title(self):
-        action = "View" if self.read_only else "Edit"
-        page_title = f"{action} Certificate Application Template"
-
-        return page_title
+        if app_type == ExportApplicationType.Types.FREE_SALE:
+            if step == CatSteps.CFS:
+                extra["schedules"] = self.object.cfs_template.schedules.all().order_by("created_at")
```
Comment (mikeylockwood) [inline web/domains/cat/views.py:267]: `.all()` does nothing here

File: web/domains/cat/views.py [hunk 5]
```diff
@@ -209,6 +308,11 @@ def get_sidebar_links(self):
         match self.object.application_type:
             case ExportApplicationType.Types.FREE_SALE:
                 app_specific: list[tuple[str, str]] = []
+
+                template: CertificateOfFreeSaleApplicationTemplate = self.object.cfs_template
+                for s_id, s in enumerate(template.schedules.all().order_by("id"), start=1):
```
Comment (mikeylockwood) [inline web/domains/cat/views.py:313]: Another :)

File: web/domains/cat/views.py [hunk 6]
```diff
@@ -269,3 +395,188 @@ def post(self, request: AuthenticatedHttpRequest, *args, **kwargs) -> HttpRespon
...
+    schedule.created_by = user
+    schedule.save()
+
+    # Copy the legislation records
+    schedule.legislations.set(legislations_to_copy)
+
+    # copy the product records
+    for product in products_to_copy:
+        product_types_to_copy = [pt for pt in product.product_type_numbers.all().order_by("pk")]
+        ingredients_to_copy = [i for i in product.active_ingredients.all().order_by("pk")]
```
Comment (mikeylockwood) [inline web/domains/cat/views.py:541]: Lots of `.all()`

File: web/templates/web/domains/cat/base_form.html [hunk 7]
```diff
@@ -31,80 +31,68 @@
...
+          {% if not read_only %}
             <ul class="menu-out flow-across">
               <li>
                 <button id="btn-submit" type="submit" class="primary-button button">
-                  {% if form.instance %}
-                    Save
-                  {% else %}
-                    Create
-                  {% endif %}
+                  {% if form.instance %} Save {% else %} Create {% endif -%}
```
Comment (mikeylockwood) [inline web/templates/web/domains/cat/base_form.html:54]: Is there a reason just the trailing whitespace is stripped? Might as well strip all of it

**Review by mikeylockwood — APPROVED** [review]

File: web/templates/web/domains/cat/base_form.html (hunk 7 above)
Comment (MattHolmes123) [reply web/templates/web/domains/cat/base_form.html:54]: Nope, I didn't even notice that i added it tbh.

Comment (MattHolmes123) [reply web/templates/web/domains/cat/base_form.html:54]: Sorted it anyway

File: web/domains/cat/views.py (hunk 5 above)
Comment (MattHolmes123) [reply web/domains/cat/views.py:313]: I love them :)

File: web/domains/cat/views.py (hunk 6 above)
Comment (MattHolmes123) [reply web/domains/cat/views.py:541]: 🤸 🤸 🤸

**Pull request merged**